internal/client/client_encrypted.go

```diff
@@ -0,0 +1,233 @@
+package client
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+	"maunium.net/go/mautrix"
+	"maunium.net/go/mautrix/crypto"
+	"maunium.net/go/mautrix/event"
+	"maunium.net/go/mautrix/id"
+
+	"github.com/matrix-org/complement/internal/must"
+)
+
+type CSAPIEncrypted struct {
+	*CSAPI
+	mautrix    *mautrix.Client
+	olmMachine *crypto.OlmMachine
+	stateStore StateStore
+	encChan    chan EncryptedEventResponse
+}
+
+type EncryptedEventResponse struct {
+	Body  string
+	Error error
+}
+
+type clientOptions struct {
+	logger crypto.Logger
+}
+
+type ClientOption func(*clientOptions)
+
+// WithLogging enables a crypto.Logger using t.Log/t.Fatal etc.
+func WithLogging(t *testing.T, csapi *CSAPI) ClientOption {
+	return func(opts *clientOptions) {
+		opts.logger = testCryptoLogger{
+			t:      t,
+			userID: csapi.UserID,
+		}
+	}
+}
+
+// MustCreateEncryptedClient creates an "encrypted client" from a CSAPI and initialises
+// the OlmMachine, fails the test if the client or OlmMachine can't be created
+func MustCreateEncryptedClient(t *testing.T, csapi *CSAPI, opts ...ClientOption) *CSAPIEncrypted {
+	defaultOptions := &clientOptions{
+		logger: emptyCryptoLogger{},
+	}
+
+	for _, opt := range opts {
+		opt(defaultOptions)
+	}
+
+	cl, err := mautrix.NewClient(csapi.BaseURL, id.UserID(csapi.UserID), csapi.AccessToken)
+	if err != nil {
+		t.Fatalf("failed to create mautrix.Client: %s", err)
+	}
+	cl.DeviceID = id.DeviceID(csapi.DeviceID)
+
+	stateStore := StateStore{storer: mautrix.NewInMemoryStore()}
+
+	// Open SQL connection
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open sql.DB: %s", err)
+	}
+
+	accountID := csapi.UserID + "-" + csapi.DeviceID
+	sqlCryptoStore := crypto.NewSQLCryptoStore(db, "sqlite3", accountID, id.DeviceID(csapi.DeviceID), []byte(csapi.DeviceID+"pickle"), defaultOptions.logger)
```

Comment:
Having spoken to tulir, we can use `NewGobStore` safely so long as we don't write to disk as you can lose data if the write fails. Please can we use that but yank out the save() code, so either copy https://github.com/mautrix/go/blob/v0.12.1/crypto/store.go#L354 and comment out the `save()` code or write to `/dev/null`. I'd much rather use an in-memory impl than pull in a SQLite dep and add more moving parts.